apis/python/src/tiledbsc/soma_slice.py

```diff
@@ -0,0 +1,124 @@
+        obs: pd.DataFrame,
+        var: pd.DataFrame,
+    ):
+        """
+        Constructs an in-memory `SOMASlice` object. This is a simple collection of obs, var, and X dataframes.
+        """
+        assert isinstance(obs, pd.DataFrame)
+        assert isinstance(var, pd.DataFrame)
+        assert "data" in X
+
+        self.obs = obs
+        self.var = var
+        self.X = X
+
+    # ----------------------------------------------------------------
+    def to_anndata(self) -> ad.AnnData:
+        """
+        Constructs an `AnnData` object from the current `SOMASlice` object.
+        """
+
+        # Find the dtype.
+        X_data = self.X["data"]
+        if isinstance(X_data, pd.DataFrame):
+            X_dtype = X_data.dtypes["value"]
+        else:
+            X_dtype = X_data.dtype
+
+        ann = ad.AnnData(obs=self.obs, var=self.var, dtype=X_dtype)
+
+        for name, data in self.X.items():
+            # X comes in from TileDB queries as a 3-column dataframe with "obs_id", "var_id", and
+            # "value".  For AnnData we need to make it a sparse matrix.
+            if isinstance(data, pd.DataFrame):
+                # Make obs_id and var_id accessible as columns.
+                data = data.reset_index()
+                data = util.X_and_ids_to_sparse_matrix(
+                    data,
+                    "obs_id",  # row_dim_name
+                    "var_id",  # col_dim_name
+                    "value",  # attr_name
+                    self.obs.index,
+                    self.var.index,
+                )
+            # We use AnnData as our in-memory storage. For SOMAs, all X layers are arrays within the
+            # soma.X group; for AnnData, the 'data' layer is ann.X and all the others are in
+            # ann.layers.
+            if name == "data":
+                ann.X = data
+            else:
+                ann.layers[name] = data
+
+        return ann
+
+    # ----------------------------------------------------------------
+    @classmethod
+    def concat(cls, soma_slices):
+        """
+        Concatenates multiple `SOMASlice` objects into a single one. Implemented using `AnnData`'s
+        `concat`.
+        """
+
+        if len(soma_slices) == 0:
+            return None
+
+        # Check column names for each dataframe-type are the same
+        slice0 = soma_slices[0]
+        for i, slicei in enumerate(soma_slices):
+            if i == 0:
+                continue
+            # This works in Python -- not just a reference/pointer compare but a contents-compare :)
+            assert list(slicei.X.keys()) == list(slice0.X.keys())
+            assert list(slicei.obs.keys()) == list(slice0.obs.keys())
+            assert list(slicei.var.keys()) == list(slice0.var.keys())
+
+        # Use AnnData.concat.
+        # TODO: try to remove this dependency.
+        anns = [soma_slice.to_anndata() for soma_slice in soma_slices]
+        annc = ad.concat(anns)
```

Comment:
this is an inner join on the second axis, which is not what you want....it will result in data being dropped.

It should read:  `ad.concat(anns, join="outter")`

Reply:
done on a commit by @bkmartinjr :)